WebGLRenderingContextBase::readPixels should early reject FLOAT_32_UNSIGNED_INT_24_8_REV type

kkinnunen-apple opened the pull request:
#### 40438efc37ff8bdac5deacad93630826b8d0cb63
<pre>
WebGLRenderingContextBase::readPixels should early reject FLOAT_32_UNSIGNED_INT_24_8_REV type
<a href="https://bugs.webkit.org/show_bug.cgi?id=244290">https://bugs.webkit.org/show_bug.cgi?id=244290</a>
rdar://problem/99081375

Reviewed by Kenneth Russell.

Rename validateArrayBufferType to validateTypeAndArrayBufferType to better
reflect what the function does.

Add validation for type == FLOAT_32_UNSIGNED_INT_24_8_REV for read pixels.
It should always fail with INVALID_ENUM, regardless whether pixels are passed or not.

For texImage functions, the type != null case produces INVALID_OPERATION.

Tested by:
webgl/2.0.y/conformance/programs/program-test.html

TestExpectations is not yet changed as the test result is
Pass/Fail and currently another hunk of the test fails.

* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::readPixels):
(WebCore::WebGLRenderingContextBase::validateTypeAndArrayBufferType):
(WebCore::WebGLRenderingContextBase::validateTexFuncData):
WebCore::WebGLRenderingContextBase::validateArrayBufferType): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:

Canonical link: <a href="https://commits.webkit.org/255165@main">https://commits.webkit.org/255165@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9aa39b32b1d698b06ba3c1ccb53dd01829e2337e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91624 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/833 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/22226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101322 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/161402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95629 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/829 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83941 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/97657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97282 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27444 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/82410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/22226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/70481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35746 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/22226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33501 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/22226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3587 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37342 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39247 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/22226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

kenrussell reviewed (approved):
Looks good to me. Per the PR description, expecting a roll-forward of the WebGL conformance suite into WebKit soon.

pedro-ximenez reviewed (approved):
:shipit:

kkinnunen-apple committed:
WebGLRenderingContextBase::readPixels should early reject FLOAT_32_UNSIGNED_INT_24_8_REV type
https://bugs.webkit.org/show_bug.cgi?id=244290
rdar://problem/99081375

Reviewed by Kenneth Russell.

Rename validateArrayBufferType to validateTypeAndArrayBufferType to better
reflect what the function does.

Add validation for type == FLOAT_32_UNSIGNED_INT_24_8_REV for read pixels.
It should always fail with INVALID_ENUM, regardless whether pixels are passed or not.

For texImage functions, the type != null case produces INVALID_OPERATION.

Tested by:
webgl/2.0.y/conformance/programs/program-test.html

TestExpectations is not yet changed as the test result is
Pass/Fail and currently another hunk of the test fails.

* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::readPixels):
(WebCore::WebGLRenderingContextBase::validateTypeAndArrayBufferType):
(WebCore::WebGLRenderingContextBase::validateTexFuncData):
WebCore::WebGLRenderingContextBase::validateArrayBufferType): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:

Canonical link: https://commits.webkit.org/255165@main
